File: array/problem152/problem152.py
from typing import List


class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        cur_max = cur_min = best = nums[0]

        for x in nums[1:]:
            prev_max = cur_max
            cur_max = max(x, prev_max * x, cur_min * x)
            cur_min = min(x, prev_max * x, cur_min * x)
            if cur_max > best:
                best = cur_max

        return best

File: array/problem152/problem152.cpp
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int curMax = nums[0];
        int curMin = nums[0];
        int best   = nums[0];

        for (int i = 1; i < static_cast<int>(nums.size()); ++i) {
            int x = nums[i];
            int prevMax = curMax;
            curMax = max({x, prevMax * x, curMin * x});
            curMin = min({x, prevMax * x, curMin * x});
            best   = max(best, curMax);
        }

        return best;
    }
};
